File: leetcode/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        LinkedList<TreeNode> stack = new LinkedList<TreeNode>();
        TreeNode previous = null;
        
        while(root != null || stack.size() != 0) {
            while(root != null) {
                stack.addFirst(root);
                root = root.left;
            }
            root = stack.removeFirst();
            if(previous != null && root.val <= previous.val)
                return false;
            previous = root;
            root = root.right;
        }
        return true;
    }
}

File: leetcode/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode], min_val=-math.inf, max_val=math.inf) -> bool:
        if not root:
            return True
        
        if root.val >= max_val or root.val <= min_val:
            return False
        
        left = self.isValidBST(root.left, min_val, min(max_val, root.val))
        right = self.isValidBST(root.right, max(min_val, root.val), max_val)
        return left and right
